auth/range_perm_cache.go

```diff
@@ -0,0 +1,160 @@
+	return len(slice)
+}
+
+func (slice RangePermSlice) Less(i, j int) bool {
+	return slice[i].begin < slice[j].begin
+}
+
+func (slice RangePermSlice) Swap(i, j int) {
+	slice[i], slice[j] = slice[j], slice[i]
+}
+
+type unifiedRangePermissions struct {
+	// readPerms[i] and readPerms[j] (i != j) don't overlap
+	readPerms []*rangePerm
+	// writePerms[i] and writePerms[j] (i != j) don't overlap, too
+	writePerms []*rangePerm
+}
+
+func unifyPerms(perms []*rangePerm) []*rangePerm {
+	ret := make([]*rangePerm, 0)
+	sort.Sort(RangePermSlice(perms))
+
+	i := 0
+	for i < len(perms) {
+		begin := i
+		for i < len(perms) && perms[i].end >= perms[i+1].begin {
+			i++
+		}
+
+		if i == begin {
+			ret = append(ret, &rangePerm{begin: perms[i].begin, end: perms[i].end})
+			i++
+		} else {
+			ret = append(ret, &rangePerm{begin: perms[begin].begin, end: perms[i].end})
+		}
+	}
+
+	return ret
+}
+
+func (as *authStore) makeUnifiedPerms(tx backend.BatchTx, userName string) *unifiedRangePermissions {
+	_, vs := tx.UnsafeRange(authUsersBucketName, []byte(userName), nil, 0)
+	if len(vs) != 1 {
+		plog.Errorf("invalid user name %s", userName)
+		return nil
+	}
+
+	user := &authpb.User{}
+	err := user.Unmarshal(vs[0])
+	if err != nil {
+		plog.Errorf("failed to unmarshal user struct (name: %s): %s", userName, err)
+		return nil
+	}
+
+	var readPerms, writePerms []*rangePerm
+
+	for _, roleName := range user.Roles {
+		_, vs := tx.UnsafeRange(authRolesBucketName, []byte(roleName), nil, 0)
+		if len(vs) != 1 {
+			plog.Errorf("invalid role name %s", roleName)
+			return nil
+		}
+
+		role := &authpb.Role{}
+		err := role.Unmarshal(vs[0])
+		if err != nil {
+			plog.Errorf("failed to unmarshal a role %s: %s", roleName, err)
+			return nil
+		}
+
+		for _, perm := range role.KeyPermission {
+			if len(perm.RangeEnd) == 0 {
+				continue
+			}
+
+			if perm.PermType == authpb.READWRITE || perm.PermType == authpb.READ {
+				readPerms = append(readPerms, &rangePerm{begin: string(perm.Key), end: string(perm.RangeEnd)})
+			}
+
+			if perm.PermType == authpb.READWRITE || perm.PermType == authpb.WRITE {
+				writePerms = append(writePerms, &rangePerm{begin: string(perm.Key), end: string(perm.RangeEnd)})
+			}
+		}
+	}
+
+	return &unifiedRangePermissions{readPerms: unifyPerms(readPerms), writePerms: unifyPerms(writePerms)}
+}
+
+func checkCachedPerm(cachedPerms *unifiedRangePermissions, userName string, key, rangeEnd string, write, read bool) bool {
+	var perms []*rangePerm
+
+	if write {
+		perms = cachedPerms.writePerms
+	} else {
+		perms = cachedPerms.readPerms
+	}
+
+	for _, perm := range perms {
+		if strings.Compare(perm.begin, key) <= 0 && strings.Compare(rangeEnd, perm.end) <= 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (as *authStore) isRangeOpPermitted(tx backend.BatchTx, userName string, key, rangeEnd string, write, read bool) bool {
+	// assumption: tx is Lock()ed
+	_, ok := as.rangePermCache[userName]
+	if !ok {
```

Comment:
let's do the ok case first. it is shorter. so we have less stuff in the indented code block.